### PR TITLE
Re-implement ReqWrap and more modules in C, and rework validated struct

### PR DESCRIPTION
**Pull request by yichoi**
* Re-implement ReqWrap in C
* Re-implement Dns, Fs, Tcp in C
* Rework validated struct
  - Make private implementation (`iotjs_classname_t##_impl_t`) as its first member
  - To avoid extra work to find real starting address of the struct,
  - And to add more ASSERTs in Wrappers for safer execution

### CODE REVIEW

File: src/iotjs_reqwrap.h [hunk 1]
```diff
@@ -22,59 +22,38 @@
...
 
-  // To retrieve javascript callback funciton object.
-  const iotjs_jval_t* jcallback();
 
-  // To retrieve pointer to uv request.
-  T* req();
+void iotjs_reqwrap_initialize(iotjs_reqwrap_t* reqwrap,
+                              const iotjs_jval_t* jcallback,
+                              uv_req_t* request,
+                              void* wrap);
```
Comment (nova0821) [inline src/iotjs_reqwrap.h]: I think this line should be same line with ``uv_req_t* request, ``.

According to our coding style guideline - ``Return type should be on the same line as function name and parameters if it fits. If not, break between them aligned with the first argument.``

I think this rule reference to google's style guideline - https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions

It shows some example, like below:

```c
ReturnType ClassName::ReallyLongFunctionName(Type par_name1, Type par_name2,
                                             Type par_name3) {
  DoSomething();
  ...
}
```

File: src/module/iotjs_module_fs.c [hunk 2]
```diff
@@ -200,15 +275,12 @@ JHANDLER_FUNCTION(Read) {
     return;
   }
 
-  uv_buf_t uvbuf = uv_buf_init(reinterpret_cast<char*>(data + offset),
-                               length);
+  uv_buf_t uvbuf = uv_buf_init((char*)(data + offset), length);
```
Comment (glistening) [inline src/module/iotjs_module_fs.c]: Do we need char* casting here?

File: src/module/iotjs_module_fs.c [hunk 3]
```diff
@@ -242,15 +314,12 @@ JHANDLER_FUNCTION(Write) {
...
```
Comment (glistening) [inline src/module/iotjs_module_fs.c]: ditto

**Comment by yichoi** [comment]
fixed @nova0821 @glistening

**Comment by hs0225** [comment]
LGTM

**Comment by nova0821** [comment]
LGTM

**Pull request merged**